src/libcore/iter.rs

```diff
@@ -1312,10 +1320,15 @@ pub trait Iterator {
         let mut other = other.into_iter();
 
         loop {
-            match (self.next(), other.next()) {
-                (None, None) => return true,
-                (None, _) | (_, None) => return false,
-                (Some(x), Some(y)) => if x != y { return false },
+            let (l, r) = (self.next(), other.next());
+            return if l.is_none() && r.is_none() {
+                true
+            } else if l.is_none() || r.is_none() {
+                false
+            } else if PartialEq::ne(&l.unwrap(), &r.unwrap()) {
```

Comment:
Ah yeah I was thinking along the lines of something like `if let Some(x) = ... { } else { ... }` which in theory shouldn't suffer the tuple-in-match problems?